03 SFC Login/README.md

```diff
@@ -0,0 +1,1242 @@
+    ],
+    password: [
+      { validator: Validators.required },
+    ],
+  },
+};
+
+export const validations = createFormValidation(validationConstraints);
+
+```
+
+- Create `LoginError` model:
+
+### ./src/pages/login/viewModel.ts
+
+```diff
++ import { FieldValidationResult } from 'lc-form-validation';
+
+interface LoginEntity {
+  login: string;
+  password: string;
+};
+
+const createEmptyLoginEntity = (): LoginEntity => ({
+  login: '',
+  password: '',
+});
+
++ interface LoginError {
++   login: FieldValidationResult;
++   password: FieldValidationResult;
++ };
+
++ const createEmptyLoginError = (): LoginError => ({
++   login: {
++     key: 'login',
++     succeeded: true,
++     errorMessage: '',
++     type: '',
++   },
++   password: {
++     key: 'password',
++     succeeded: true,
++     errorMessage: '',
++     type: '',
++   },
++ });
+
+- export { LoginEntity, createEmptyLoginEntity };
++ export { LoginEntity, createEmptyLoginEntity, LoginError, createEmptyLoginError };
+
+```
+
+- Update `PageContainer`:
+
+### ./src/pages/login/PageContainer.vue
+```diff
+<template>
+  <login-page
+    :login-entity="loginEntity"
++   :login-error="loginError"
+    :update-login="updateLogin"
+    :login-request="loginRequest"
+  />
+</template>
+
+<script lang="ts">
+import Vue from 'vue';
+import { loginRequest } from '../../rest-api/api/login';
+- import { LoginEntity, createEmptyLoginEntity } from './viewModel';
++ import { LoginEntity, createEmptyLoginEntity, LoginError, createEmptyLoginError } from './viewModel';
+import { mapLoginEntityVmToModel } from './mappers';
++ import { validations } from './validations';
+import LoginPage from './Page.vue';
+
+export default Vue.extend({  
+  name: 'PageContainer',
+  components: {
+    LoginPage,
+  },
+  data() {
+    return {
+      loginEntity: createEmptyLoginEntity(),
++     loginError: createEmptyLoginError(),
+    };
+  },  
+  methods: {
+-   updateLogin(login: string, password: string) {
++   updateLogin(field: string, value: string) {
+-     this.loginEntity = {
+-       login,
+-       password,
+-     };
++     this.loginEntity = {
++       ...this.loginEntity,
++       [field]: value,
++     };
+
++    validations.validateField(this.loginEntity, field, value)
++     .then(fieldValidationResult => {
++       this.loginError = {
++         ...this.loginError,
++         [field]: fieldValidationResult,
++       };
++     })
++     .catch(error => console.log(error));
+    },
+    loginRequest() {
++     validations.validateForm(this.loginEntity)
++       .then(formValidationResult => {
++         if (formValidationResult.succeeded) {
+            const loginEntityModel = mapLoginEntityVmToModel(this.loginEntity);
+            loginRequest(loginEntityModel)
+              .then(() => {
+                this.$router.push('/recipe');
+              })
+              .catch(error => console.log(error));
++         } else {
++           for (const obj in this.loginEntity) {
```

Comment:
With `lc-form-validation` latest version we could do:

```javascript
...
} else {
  this.loginError = formValidationResult.fieldErrors;
}
```

Reply:
Perfect. I updated in 03, 04 and 05 samples.